File: android-sdk/src/main/java/com/gist/ads/sdk/ui/AdTagBridgeWebView.kt
package com.gist.ads.sdk.ui

import android.content.Intent
import android.net.Uri
import android.os.Message
import android.view.ViewGroup
import android.webkit.JavascriptInterface
import android.webkit.WebChromeClient
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.viewinterop.AndroidView

/**
 * WebView that loads the bootstrap HTML built by [com.gist.ads.sdk.utils.DisplayAdBootstrapHTML]
 * or [com.gist.ads.sdk.utils.SearchAdBootstrapHTML] and bridges its `adRendered`/passback signals
 * back to native state via a [JavascriptInterface]. Shared by both display and search ads --
 * it's fully generic, just loading whatever HTML/baseURL it's given and bridging the two message
 * names. Also implements [WebChromeClient.onCreateWindow] to intercept the ad tag's
 * `target="_blank"` links (confirmed by reading adtag.js: ad links always render with
 * `target="_blank"`), which `WebView` would otherwise silently drop without this override.
 */
private class PrtagBridge(
    var onAdRendered: ((Double) -> Unit)?,
    var onNoFill: (() -> Unit)?
) {
    @JavascriptInterface
    fun onAdRendered(height: Double) {
        onAdRendered?.invoke(height)
    }

    @JavascriptInterface
    fun onNoFill() {
        onNoFill?.invoke()
    }
}

/**
 * Intercepts `target="_blank"`/`window.open()` navigations from the embedded
 * ad tag by creating a temporary, never-attached `WebView` whose first
 * navigation is captured and handed off natively instead of actually being
 * loaded -- the standard Android pattern for `onCreateWindow`.
 */
private class AdTagWebChromeClient(
    private val onAdClicked: (String) -> Unit
) : WebChromeClient() {
    override fun onCreateWindow(
        view: WebView,
        isDialog: Boolean,
        isUserGesture: Boolean,
        resultMsg: Message
    ): Boolean {
        val transientWebView = WebView(view.context)
        transientWebView.webViewClient = object : WebViewClient() {
            override fun shouldOverrideUrlLoading(v: WebView, request: WebResourceRequest): Boolean {
                onAdClicked(request.url.toString())
                return true
            }

            @Suppress("DEPRECATION", "OVERRIDE_DEPRECATION")
            override fun shouldOverrideUrlLoading(v: WebView, url: String): Boolean {
                onAdClicked(url)
                return true
            }
        }

        val transport = resultMsg.obj as? WebView.WebViewTransport
        transport?.webView = transientWebView
        resultMsg.sendToTarget()
        return true
    }
}

private class AdTagWebViewClient(
    private val allowedHost: String?,
    private val onAdClicked: (String) -> Unit,
    private val onLoadFailure: (String) -> Unit
) : WebViewClient() {
    override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest): Boolean {
        val url = request.url
        val scheme = url.scheme?.lowercase()
        if (scheme != "http" && scheme != "https") {
            return false
        }
        // Same-frame http(s) navigations shouldn't normally happen (ad
        // links use target="_blank", handled by AdTagWebChromeClient
        // above), but treat any as an ad click defensively rather than
        // letting the WebView navigate away from the bootstrap document.
        if (url.host != null && url.host == allowedHost) {
            return false
        }
        onAdClicked(url.toString())
        return true
    }

    override fun onReceivedError(
        view: WebView,
        request: WebResourceRequest,
        error: android.webkit.WebResourceError
    ) {
        if (request.isForMainFrame) {
            onLoadFailure(error.description?.toString() ?: "WebView load failed")
        }
    }
}

/**
 * Composable embedding the real ad tag for a single ad slot (display or search).
 *
 * @param html Bootstrap HTML document to load (see [com.gist.ads.sdk.utils.DisplayAdBootstrapHTML]
 *   / [com.gist.ads.sdk.utils.SearchAdBootstrapHTML]).
 * @param baseUrl Base URL used for [WebView.loadDataWithBaseURL] -- must be a real origin (not
 *   null) matching the ad tag script's host, both so the tag's own cross-origin fetches see a
 *   proper `Origin` header and so relative resource resolution behaves like a real embed.
 * @param theme Theme preference - "light", "dark", or "system".
 * @param onAdRendered Invoked when the ad tag reports `adRendered`, with the measured height in px.
 * @param onNoFill Invoked when the ad tag invokes its passback function.
 * @param onLoadFailure Invoked on a native WebView-level load failure.
 * @param onAdClicked Invoked when the user clicks the ad; if null, opens the URL in the default browser.
 */
// PrtagBridge's methods are correctly annotated with @JavascriptInterface (see above); lint's
// UAST type resolution loses the concrete type across the remember{}/apply{} lambda boundaries
// here and misreports the type as the generic parameter name "T" -- a known resolver limitation,
// not a real missing-annotation issue.
@Suppress("JavascriptInterface")
@Composable
fun AdTagBridgeWebView(
    html: String,
    baseUrl: String,
    modifier: Modifier = Modifier,
    theme: String = "system",
    onAdRendered: ((Double) -> Unit)? = null,
    onNoFill: (() -> Unit)? = null,
    onLoadFailure: ((String) -> Unit)? = null,
    onAdClicked: ((String) -> Unit)? = null
) {
    val bridge: PrtagBridge = remember { PrtagBridge(onAdRendered, onNoFill) }
    bridge.onAdRendered = onAdRendered
    bridge.onNoFill = onNoFill

    fun handleExternalUrl(context: android.content.Context, url: String) {
        if (onAdClicked != null) {
            onAdClicked(url)
        } else {
            val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
            context.startActivity(intent)
        }
    }

    val allowedHost = remember(baseUrl) { Uri.parse(baseUrl).host }

    AndroidView(
        modifier = modifier,
        factory = { context ->
            WebView(context).apply {
                layoutParams = ViewGroup.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT
                )

                settings.apply {
                    javaScriptEnabled = true
                    domStorageEnabled = true
                    loadWithOverviewMode = true
                    useWideViewPort = true
                    setSupportZoom(false)
                    javaScriptCanOpenWindowsAutomatically = true
                    setSupportMultipleWindows(true)
                }

                setBackgroundColor(android.graphics.Color.TRANSPARENT)
                addJavascriptInterface(bridge, "prtagBridge")

                webViewClient = AdTagWebViewClient(
                    allowedHost = allowedHost,
                    onAdClicked = { url -> handleExternalUrl(context, url) },
                    onLoadFailure = { message -> onLoadFailure?.invoke(message) }
                )
                webChromeClient = AdTagWebChromeClient(
                    onAdClicked = { url -> handleExternalUrl(context, url) }
                )

                applyThemeToWebView(this, theme)
                loadDataWithBaseURL(baseUrl, html, "text/html", "UTF-8", null)
            }
        },
        update = { webView ->
            applyThemeToWebView(webView, theme)
        },
        onRelease = { webView ->
            webView.removeJavascriptInterface("prtagBridge")
            webView.destroy()
        }
    )
}

File: swift-sdk/Sources/GistAdsSDK/Views/AdTagBridgeWebView.swift
//
//  AdTagBridgeWebView.swift
//  GistAdsSDK
//
//  WebView that loads the bootstrap HTML built by DisplayAdBootstrapHTML.swift
//  or SearchAdBootstrapHTML.swift and bridges its `adRendered`/passback
//  signals back to native state via WKScriptMessageHandler. Shared by both
//  `GistDisplayAdControl` and `GistAdControl` -- it's a generic embed shell
//  that just loads whatever HTML/baseURL it's given and bridges the two
//  message names; it has no knowledge of display vs. search. Also implements
//  WKUIDelegate.createWebViewWith to intercept the ad tag's target="_blank"
//  links (confirmed by reading adtag.js: ad links always render with
//  target="_blank"), which WKWebView would otherwise silently no-op on
//  without this delegate.
//

import SwiftUI
import WebKit

#if os(iOS)
import UIKit
#elseif os(macOS)
import AppKit
#endif

/// Message names registered on `WKUserContentController` for the bridge --
/// see the matching `window.webkit.messageHandlers.*` calls in
/// `DisplayAdBootstrapHTML` / `SearchAdBootstrapHTML`.
enum AdTagBridgeMessage {
    static let adRendered = "prtagAdRendered"
    static let noFill = "prtagNoFill"
}

/// Coordinator holding the bridge/navigation/UI delegate for a single
/// `AdTagBridgeWebView`.
final class AdTagBridgeCoordinator: NSObject, WKScriptMessageHandler, WKNavigationDelegate, WKUIDelegate {
    var onAdRendered: ((Double) -> Void)?
    var onNoFill: (() -> Void)?
    var onLoadFailure: ((String) -> Void)?
    var onAdClicked: ((URL) -> Void)?

    /// Tracks the last HTML string loaded, so `updateUIView`/`updateNSView`
    /// only reload the WebView when the bootstrap document actually changes.
    var lastLoadedHTML: String?

    // MARK: - WKScriptMessageHandler

    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        switch message.name {
        case AdTagBridgeMessage.adRendered:
            let height = (message.body as? NSNumber)?.doubleValue ?? 0
            onAdRendered?(height)
        case AdTagBridgeMessage.noFill:
            onNoFill?()
        default:
            break
        }
    }

    // MARK: - WKUIDelegate

    /// Intercepts `target="_blank"`/`window.open()` navigations (which is
    /// how every ad link renders -- see file header). Returning `nil`
    /// without creating a child WKWebView is Apple's documented pattern for
    /// handling these natively instead of silently dropping them.
    func webView(
        _ webView: WKWebView,
        createWebViewWith configuration: WKWebViewConfiguration,
        for navigationAction: WKNavigationAction,
        windowFeatures: WKWindowFeatures
    ) -> WKWebView? {
        if let url = navigationAction.request.url {
            handleExternalURL(url)
        }
        return nil
    }

    // MARK: - WKNavigationDelegate

    func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction, decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {
        guard let url = navigationAction.request.url else {
            decisionHandler(.allow)
            return
        }

        let scheme = url.scheme?.lowercased() ?? ""
        if scheme != "http" && scheme != "https" {
            decisionHandler(.allow)
            return
        }

        // Only main-frame navigations matter here: this covers both the
        // synthetic navigation `loadHTMLString(_:baseURL:)` generates for
        // its `baseURL` (which must be allowlisted or WebKit aborts the
        // load) and any same-frame ad click that -- unlike the tag's usual
        // target="_blank" links -- didn't go through `createWebViewWith`.
        guard navigationAction.targetFrame?.isMainFrame == true else {
            decisionHandler(.allow)
            return
        }

        if APIConstants.isAllowedAdDomain(url) {
            decisionHandler(.allow)
            return
        }

        decisionHandler(.cancel)
        handleExternalURL(url)
    }

    func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
        onLoadFailure?(error.localizedDescription)
    }

    func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
        onLoadFailure?(error.localizedDescription)
    }

    private func handleExternalURL(_ url: URL) {
        if let onAdClicked = onAdClicked {
            DispatchQueue.main.async {
                onAdClicked(url)
            }
        } else {
            DispatchQueue.main.async {
                #if os(iOS)
                UIApplication.shared.open(url, options: [:], completionHandler: nil)
                #elseif os(macOS)
                NSWorkspace.shared.open(url)
                #endif
            }
        }
    }
}

private func makeConfiguredAdTagWebView(coordinator: AdTagBridgeCoordinator, isIOS: Bool) -> WKWebView {
    let configuration = WKWebViewConfiguration()
    configuration.userContentController.add(coordinator, name: AdTagBridgeMessage.adRendered)
    configuration.userContentController.add(coordinator, name: AdTagBridgeMessage.noFill)

    let webView = WKWebView(frame: .zero, configuration: configuration)
    webView.navigationDelegate = coordinator
    webView.uiDelegate = coordinator
    configureWebView(webView, isIOS: isIOS)
    return webView
}

#if os(iOS)

/// SwiftUI wrapper embedding the real ad tag for a single ad slot (iOS).
struct AdTagBridgeWebView: UIViewRepresentable {
    let html: String
    let baseURLString: String
    var theme: String = "system"
    var onAdRendered: ((Double) -> Void)?
    var onNoFill: (() -> Void)?
    var onLoadFailure: ((String) -> Void)?
    var onAdClicked: ((URL) -> Void)?

    private var baseURL: URL {
        URL(string: baseURLString) ?? URL(string: "about:blank")!
    }

    func makeCoordinator() -> AdTagBridgeCoordinator {
        let coordinator = AdTagBridgeCoordinator()
        coordinator.onAdRendered = onAdRendered
        coordinator.onNoFill = onNoFill
        coordinator.onLoadFailure = onLoadFailure
        coordinator.onAdClicked = onAdClicked
        return coordinator
    }

    func makeUIView(context: Context) -> WKWebView {
        let webView = makeConfiguredAdTagWebView(coordinator: context.coordinator, isIOS: true)
        applyThemeToWebView(webView, theme: theme)
        webView.loadHTMLString(html, baseURL: baseURL)
        context.coordinator.lastLoadedHTML = html
        return webView
    }

    func updateUIView(_ webView: WKWebView, context: Context) {
        context.coordinator.onAdRendered = onAdRendered
        context.coordinator.onNoFill = onNoFill
        context.coordinator.onLoadFailure = onLoadFailure
        context.coordinator.onAdClicked = onAdClicked
        applyThemeToWebView(webView, theme: theme)

        if context.coordinator.lastLoadedHTML != html {
            context.coordinator.lastLoadedHTML = html
            webView.loadHTMLString(html, baseURL: baseURL)
        }
    }

    static func dismantleUIView(_ webView: WKWebView, coordinator: AdTagBridgeCoordinator) {
        webView.configuration.userContentController.removeScriptMessageHandler(forName: AdTagBridgeMessage.adRendered)
        webView.configuration.userContentController.removeScriptMessageHandler(forName: AdTagBridgeMessage.noFill)
    }
}

#elseif os(macOS)

/// SwiftUI wrapper embedding the real ad tag for a single ad slot (macOS).
struct AdTagBridgeWebView: NSViewRepresentable {
    let html: String
    let baseURLString: String
    var theme: String = "system"
    var onAdRendered: ((Double) -> Void)?
    var onNoFill: (() -> Void)?
    var onLoadFailure: ((String) -> Void)?
    var onAdClicked: ((URL) -> Void)?

    private var baseURL: URL {
        URL(string: baseURLString) ?? URL(string: "about:blank")!
    }

    func makeCoordinator() -> AdTagBridgeCoordinator {
        let coordinator = AdTagBridgeCoordinator()
        coordinator.onAdRendered = onAdRendered
        coordinator.onNoFill = onNoFill
        coordinator.onLoadFailure = onLoadFailure
        coordinator.onAdClicked = onAdClicked
        return coordinator
    }

    func makeNSView(context: Context) -> WKWebView {
        let webView = makeConfiguredAdTagWebView(coordinator: context.coordinator, isIOS: false)
        applyThemeToWebView(webView, theme: theme)
        webView.loadHTMLString(html, baseURL: baseURL)
        context.coordinator.lastLoadedHTML = html
        return webView
    }

    func updateNSView(_ webView: WKWebView, context: Context) {
        context.coordinator.onAdRendered = onAdRendered
        context.coordinator.onNoFill = onNoFill
        context.coordinator.onLoadFailure = onLoadFailure
        context.coordinator.onAdClicked = onAdClicked
        applyThemeToWebView(webView, theme: theme)

        if context.coordinator.lastLoadedHTML != html {
            context.coordinator.lastLoadedHTML = html
            webView.loadHTMLString(html, baseURL: baseURL)
        }
    }

    static func dismantleNSView(_ webView: WKWebView, coordinator: AdTagBridgeCoordinator) {
        webView.configuration.userContentController.removeScriptMessageHandler(forName: AdTagBridgeMessage.adRendered)
        webView.configuration.userContentController.removeScriptMessageHandler(forName: AdTagBridgeMessage.noFill)
    }
}

#endif
